### PR TITLE
Adding new rule no-side-effects

### LINKED ISSUE #34: Add no-side-effects rule
For tree-shaking purposes, we want to set the `sideEffects` flag in package.json to `false`.  Various bundlers use this property to work out if the exports can be tree-shaken, eliminating unused code.  But we need to make sure our libraries etc are actually side-effect-free.

Effectively, we should ensure no side effects can occur at the main module-level.  No functions called (other than some special cases), no non-constant initializations, etc.

As an additional property, we need to configure what function names can be called during module initialization that have no side effects.  The default should be:
* `debug`

### PR DESCRIPTION
**Pull request by jpolavar**
Closes #34

### CODE REVIEW

**Comment by jpolavar** [comment]
Screenshot
tested on eslint-config with the errors

File: src/no-side-effects.spec.ts [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+              const jsonSchemaValidator = new Ajv({ allErrors: true }).compile(schema);`,
+        errors: [{ message: 'No side effects can occur at the main module-level' }],
+        filename: 'src/side-effects.ts',
+      },
+      {
+        code: `import { strict as assert } from 'node:assert';
+                import http from 'node:http';
+                import debug from 'debug';
+                
+                import { getConfiguration, logger, root, service } from '@checkdigit/serve-runtime';
```
Comment (carlansley) [inline src/no-side-effects.spec.ts:127]: this is a public repo, we shouldn't have examples that make use of internal/private modules

File: src/no-side-effects.spec.ts [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+                
+                const log = debug('mngs:local');
```
Comment (carlansley) [inline src/no-side-effects.spec.ts:129]: same here

File: src/no-side-effects.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+/*
+ * Copyright (c) 2022-2024 Check Digit, LLC
+ *
+ * This code is licensed under the MIT license (see LICENSE.txt for details).
+ */
+
+import type { Rule } from 'eslint';
+
+const EXCLUDED_IDENTIFIERS = ['assert', 'debug', 'log'];
```
Comment (carlansley) [inline src/no-side-effects.ts:11]: this is ok as a default, but should be configurable.

File: src/no-side-effects.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+  create(context) {
+    return {
+      Program(node) {
+        const fileName = context.filename.split('src/')[1] ?? context.filename;
+        if (fileName.endsWith('spec.ts') || fileName.endsWith('test.ts') || fileName.endsWith('test.js')) {
+          return;
+        }
+        node.body.forEach((statement) => {
+          if (
+            (statement.type === 'ExpressionStatement' && statement.expression.type === 'AwaitExpression') ||
```
Comment (carlansley) [inline src/no-side-effects.ts:30]: this needs to be broken up somehow so its understandable.  a series of functions perhaps, not sure.  but this is too hard to parse.

**Review by carlansley — CHANGES_REQUESTED** [review]

File: src/no-side-effects.ts (hunk 3 above)
Comment (carlansley) [reply src/no-side-effects.ts:12]: this is still not a part of the rule configuration?

File: src/no-side-effects.ts [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+ *
+ * This code is licensed under the MIT license (see LICENSE.txt for details).
+ */
+
+import type { Node } from 'estree';
+import type { Rule } from 'eslint';
+
+const EXCLUDED_IDENTIFIERS = ['assert', 'debug', 'log'];
+
+function isTestFile(fileName: string): boolean {
```
Comment (carlansley) [inline src/no-side-effects.ts:14]: there is no need for this in the rule itself.  whether or not a rule runs against test files is part of eslint-config.

File: src/no-side-effects.spec.ts [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+ */
+
+import { RuleTester } from '@typescript-eslint/rule-tester';
+import { describe } from '@jest/globals';
+
+import rule from './no-side-effects';
+
+describe('no-side-effects', () => {
+  const ruleTester = new RuleTester();
+  // @ts-expect-error, workaround for the type definition
```
Comment (carlansley) [inline src/no-side-effects.spec.ts:16]: are we sure we need this?  seems concerning

**Review by carlansley — CHANGES_REQUESTED** [review]

File: src/no-side-effects.spec.ts (hunk 6 above)
Comment (jpolavar) [reply src/no-side-effects.spec.ts:16]: RuleTester imported from eslint doesn't parse the test code properly and fails but RuleTester from @typescript-eslint/rule-tester works as expected

**Review by carlansley — CHANGES_REQUESTED** [review]
the code lgtm, but can you do a beta eslint-config with this beta and check it against some of our bigger internal repos?  I suspect there might be some false positives we need to deal with.

**Comment by jpolavar** [comment]
tested on mngs 
Screenshot 2024-07-02 at 3 02 08 PM

File: src/no-side-effects.ts [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+  return statement.type === 'ExpressionStatement' && statement.expression.type === 'AwaitExpression';
+}
+
+function isCallExpressionCalleeMemberExpression(statement: Node): boolean {
+  return (
+    statement.type === 'ExpressionStatement' &&
+    statement.expression.type === 'CallExpression' &&
+    statement.expression.callee.type === 'MemberExpression' &&
+    statement.expression.callee.object.type === 'Identifier' &&
+    statement.expression.callee.object.name !== 'assert'
```
Comment (le-cong) [inline src/no-side-effects.ts:26]: should this be handled similarly using configuration?

File: src/no-side-effects.spec.ts [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+                
+                // wait until the server is listening
+                await new Promise((resolve) => {
+                  server.on('listening', resolve);
+                });
+                
+                log(
+                  \`started \${unresolvedConfiguration.name} on http://localhost:\${unresolvedConfiguration.env['PORT']}, using node.js \${process.version}\`,
+                );`,
+        errors: [
```
Comment (le-cong) [inline src/no-side-effects.spec.ts:117]: shall we split this test case (code) and maybe remove the valid statements as well so that it's easier to understand which statements triggger the errors?

File: src/no-side-effects.ts [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+
+function isVariableDeclarationAwaitExpression(statement: Node): boolean {
+  return (
+    statement.type === 'VariableDeclaration' &&
+    statement.declarations.length > 0 &&
+    statement.declarations[0]?.id !== undefined &&
+    statement.declarations[0].id.type === 'Identifier' &&
+    statement.declarations[0].init !== undefined &&
+    statement.declarations[0].init !== null &&
+    statement.declarations[0].init.type === 'AwaitExpression'
```
Comment (le-cong) [inline src/no-side-effects.ts:47]: i could be wrong, but it feels like the check for init can reuse the logic above. similarly for the following 2 variable declaration related init checking as well.

File: src/no-side-effects.ts [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+      statement.declarations[0].init.callee.object.callee.type === 'Identifier') ||
+      statement.declarations[0].init.callee.object.type === 'Identifier')
+  );
+}
+
+export default {
+  meta: {
+    type: 'problem',
+    docs: {
+      description: 'Ensure no side effects can occur at the main module-level',
```
Comment (le-cong) [inline src/no-side-effects.ts:85]: similar here, maybe just "at the module level"?

**Review by le-cong — CHANGES_REQUESTED** [review]

File: src/no-side-effects.ts (hunk 9 above)
Comment (jpolavar) [reply src/no-side-effects.ts:47]: Somehow it likes eslint traversal only if we have the above checks and can't be reused.
